fw/apm.h

```diff
@@ -79,6 +81,91 @@ typedef struct {
 	TfwHMCodeStats	*rsums;
 } TfwHMStats;
 
+/**
+ * The real size of TfwHMStats, used for memory allocation.
+ *
+ * TfwHMCodeStats is located monolithically in memory with TfwHMStats,
+ * just after it.
+ */
+static inline size_t
+tfw_hm_stats_size(int ccnt)
+{
+	return sizeof(TfwHMStats) + sizeof(TfwHMCodeStats) * ccnt;
+}
+
+/**
+ * Use it right after memory allocation.
+ *
+ * Rsums content stays unitinitialized.
+ */
+static inline void
+tfw_hm_stats_init(TfwHMStats *s, int ccnt)
+{
+	s->ccnt = ccnt;
+	s->rsums = (TfwHMCodeStats *)(s + 1);
+}
+
+static inline void
+tfw_hm_stats_clone(TfwHMStats *dest, TfwHMStats *src)
+{
+	if (!src)
+		return;
+
+	memcpy_fast(dest, src, tfw_hm_stats_size(src->ccnt));
+	dest->rsums = (TfwHMCodeStats *)(dest + 1);
+}
+
+/**
+ * The config entry @ce must be a list of HTTP codes, such as:
+ * 'some_directive 403 404 5*'.
+ *
+ * Use it right after memory allocation.
+ */
+static inline int
+tfw_hm_stats_init_from_cfg_entry(TfwHMStats *s, TfwCfgEntry *ce)
+{
+	int i, code;
+	const char *val;
+	tfw_hm_stats_init(s, ce->val_n);
+
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		BUG_ON(i >= s->ccnt);
+
+		if (tfw_cfgop_parse_http_status(val, &code)
+		    || (code > HTTP_STATUS_5XX
+			&& tfw_cfg_check_range(code, HTTP_CODE_MIN,
```

Comment:
We also check range int the `tfw_cfgop_parse_http_status` why we should do it here?

Reply:
This is some mistake probably, it was in original code, I'll fix it.